src/util/server.ml

```diff
@@ -0,0 +1,175 @@
+open Batteries
+open Jsonrpc
+
+exception Failure of Response.Error.Code.t * string
+
+type t = {
+  file: Cil.file;
+  do_analyze: Analyses.increment_data -> Cil.file -> unit;
+  input: IO.input;
+  output: unit IO.output;
+}
+
+module type Request = sig
+  val name: string
+
+  type params
+  type response
+
+  val params_of_yojson: Yojson.Safe.t -> (params, string) result
+  val response_to_yojson: response -> Yojson.Safe.t
+
+  val process: params -> t -> response
+end
+
+module Registry = struct
+  type t = (string, (module Request)) Hashtbl.t
+  let make () : t = Hashtbl.create 32
+  let register (reg: t) (module R : Request) = Hashtbl.add reg R.name (module R)
+end
+
+let registry = Registry.make ()
+
+let handle_exn id exn =
+  Response.Error.(make Code.InternalError (Printexc.to_string exn) () |> Response.error id)
+
+module ParamParser (R : Request) = struct
+  let parse params =
+    let maybe_params =
+      params
+      |> Option.map_default Message.Structured.to_json `Null
+      |> R.params_of_yojson
+    in
+    match maybe_params with
+    | Ok params -> Ok params
+    | Error err ->
+      (* This is a hack to handle cases where R.params is a primitive type like int or string. *)
+      match params with
+      | Some `List [param] -> R.params_of_yojson param |> Result.map_error (fun _ -> err)
+      | _ -> Error err
+end
+
+let handle_request (serv: t) (message: Message.either) (id: Id.t) =
+  let req = Hashtbl.find_option registry message.method_ in
+  let response = match req with
+    | Some (module R) ->
+      let module Parser = ParamParser (R) in (
+        match Parser.parse message.params with
+        | Ok params -> (
+            try
+              R.process params serv
+              |> R.response_to_yojson
+              |> Response.ok id
+            with Failure (code, msg) -> Response.Error.(make code msg () |> Response.error id))
+        | Error s -> Response.Error.(make Code.InvalidParams s () |> Response.error id))
+    | _ -> Response.Error.(make Code.MethodNotFound message.method_ () |> Response.error id)
+  in
+  Response.yojson_of_t response |> Yojson.Safe.to_string |> IO.write_line serv.output;
+  IO.flush serv.output
+
+let serve serv =
+  serv.input
+  |> IO.to_input_channel
+  |> Yojson.Safe.linestream_from_channel
+  |> Stream.iter (fun line ->
+      match line with
+      | `Json json -> (
+          try
+            let message = Message.either_of_yojson json in
+            match message.id with
+            | Some id -> handle_request serv message id
+            | _ -> () (* We just ignore notifications for now. *)
+          with Json.Of_json (s, _) -> prerr_endline s)
+      | `Exn exn -> prerr_endline (Printexc.to_string exn))
+
+let make ?(input=stdin) ?(output=stdout) file do_analyze : t = { file; do_analyze; input; output }
+
+let bind () =
+  let server = GobConfig.get_string "server" in
+  if server = "-" then None, None else (
```

Comment:
I was about to ask why this is `"-"` not the default `""`, but I get it now:
* `""` means no server mode.
* `"-"` means server mode with stdio.
* `"otherwise"` means server mode with "otherwise" as UNIX socket.

I think this might be overloading the `server` option a bit too much already. It would be clearer and more future-proof to make `server` a category and have numerous options under it like:
* `server.enabled` - boolean.
* `server.mode` - `stdio`/`unix`
* `server.unix-socket` - name of UNIX socket to use in `unix` mode.

Reply:
Good idea. Your comment also reminded me to update the option description since "Server mode" is a bit unexplanatory now.